torch/_functorch/_aot_autograd/runtime_wrappers.py

```diff
@@ -2031,6 +2031,22 @@ def _backward_impl(ctx, all_args):
                         ),
                     )
 
+                # not ideal but prevent the user from seeing a nasty traceback - See #138422
+                stack = torch._C._functorch.peek_interpreter_stack()
+                forbidden_keys = (
+                    torch._C._functorch.TransformType.Vmap,
+                    torch._C._functorch.TransformType.Grad,
+                    torch._C._functorch.TransformType.Jvp,
+                )
```

Comment:
No need to check for forbidden keys, I think if there is anything on the stack we can raise this error.

Reply:
Given the recent refactors for compiled autograd (cc @xmfan), we might want to move these checks into `_backward_prologue` instead of `_backward_impl`?

`_backward_impl` will be inlined by dynamo, so we'd need to make sure that dynamo does something reasonable with this error check (maybe it will?)

`_backward_prologue` will by hidden from dynamo, and traced inside of AOTAutograd later. So no extra machinery is needed to get this error checking working, as long as the functorch stack is still properly set by the time compiled autograd sends the graph to dynamo/AOTAutograd for tracing

Reply:
Good point, checks should be in backward_prologue.

Reply:
Moved to `backward_prologue`